# Report 
Group 5 - Product in consideration Diacetyl (mmol/gDW/h)\
Guilherme Lobo\
Karyna Lysenko\
Rodrigo Esperança

Biologia de Sistemas - Universidade do Minho\
December 2022

### Documentação
https://cobrapy.readthedocs.io/en/latest/

Note: The results of simulations will vary each time that the code is runned.

In [2]:
import cobra
from cobra.io import read_sbml_model
from cobra.flux_analysis import flux_variability_analysis as fva
import pandas
import mewpy
from mewpy.simulation import get_simulator
from mewpy.omics import ExpressionSet
from mewpy.omics import GIMME

In [5]:
model = read_sbml_model("sth_novo.xml")

#### To find out the Diacetyl metabolite id:

In [92]:
for i in model.metabolites:
    if i.name == "Diacetyl_":
        print(i.id)

diact_c
diact_e


In [93]:
diact_c=model.metabolites.get_by_id("diact_c")
#diact_c
diact_c=model.metabolites.get_by_id("diact_e")
#diact_c
DIACTt= model.reactions.get_by_id("DIACTt")
#DIACTt
casp= model.reactions.get_by_id("EX_caspep_e")
#casp
EX_diact_e= model.reactions.get_by_id("EX_diact_e")
#EX_diact_e

## Chemical defined medium (CDM)
### Function objetive: maximization of biomass production.

##### 1 a) What is the wild-type production of the compound?


In order to have a general idea of the flux of biomass_STR and of the exchange reactions we applied the summary() function and noted the biomass production is 1.042 1/h.

In [153]:
model.summary()

In [154]:
model.optimize()["EX_diact_e"]

-0.0

Conclusion: The wild-type production of the Diacetyl is 0.0 mmol/gDW/h.

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [96]:
model_fva=fva(model,fraction_of_optimum=1)
model_fva.loc["EX_diact_e"]

minimum    0.0
maximum    0.0
Name: EX_diact_e, dtype: float64

Conclusion: As the minimum and maximum are very close to each other (maximum is almost 0.0), we can conclude that the solution is robust.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [97]:
model_1c = read_sbml_model("sth_novo.xml")

In [98]:
model_1c.summary(fva=0.20)

In [99]:
model_1c.optimize()["EX_diact_e"]

-0.0

Conclusion: The Diacetyl production is still 0.0 mmol/gDW/h.

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.

In [108]:
#saving Diacetyl production, gene name, growth performance and Lactose consumption in a list
list_variants_D=[]
for i in model.genes:
    with model:
        variant=()
        objective,growth=0,0
        gene=''
        i.knock_out()
        tmp=model.optimize()
        save_diact_optimize=float(tmp["EX_diact_e"])
        save_lactose_optimize=float(tmp["EX_lcts_e"])
        gene=str(i)
        objective=model.slim_optimize()
        variant = (save_diact_optimize, gene, objective,save_lactose_optimize)
        list_variants_D.append(variant)

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

In [109]:
#ranking in order of Diacetyl production
sorted(list_variants_D, reverse=True)

[(3.785000000000082, 'CH8_0709', 0.0, -25.11),
 (3.785000000000082, 'CH8_0621', 0.0, -25.11),
 (3.784999999999968, 'CH8_1893', 0.0, -25.11),
 (3.784999999999865, 'CH8_1630', 0.0, -25.11),
 (3.7849999999998545, 'CH8_1628', 0.0, -25.11),
 (2.7850000000003092, 'CH8_2056', 0.0, -25.11),
 (1.0332692307694593, 'CH8_1333', 0.0, -25.11),
 (0.7472093023261388, 'CH8_1671', 0.0, -25.11),
 (0.7472093023261388, 'CH8_1663', 0.0, -25.11),
 (0.7472093023255915, 'CH8_1652', 0.0, -25.11),
 (0.7472093023253592, 'CH8_0641', 0.0, -25.11),
 (0.7045294117648482, 'CH8_2144', 0.0, -25.11),
 (0.4356410256411003, 'CH8_0620', 0.0, -25.11),
 (0.3481250000000213, 'CH8_0247', 0.0, -25.11),
 (0.28500000000005343, 'CH8_0368', 0.0, -25.11),
 (0.17887254901950556, 'CH8_0289', 0.0, -25.11),
 (2.0961010704922955e-13, 'CH8_0352', 0.0, -25.11),
 (-0.0, 'CH8_2151', 0.0, -25.11),
 (-0.0, 'CH8_2150', 1.039138055405147, -25.11),
 (-0.0, 'CH8_2143', 0.0, -25.11),
 (-0.0, 'CH8_2142', 0.0, -25.11),
 (-0.0, 'CH8_2141', 0.0, -25.11)

In [110]:
#ranking in order of growth performance
import numpy as np
cleanedListP = []
for y in list_variants_D:
    if not np.isnan(y[2]):
        cleanedListP.append(y)
        
def Sort(sub_li):
    sub_li.sort(key = lambda x:x[2], reverse=True)
    return sub_li
growthList=Sort(cleanedListP)
growthList

[(-0.0, 'CH8_0712', 1.0422533807372427, -25.11),
 (-0.0, 'CH8_0714', 1.0422533807372427, -25.11),
 (-0.0, 'CH8_0648', 1.042253380736947, -25.11),
 (-0.0, 'CH8_0654', 1.042253380736947, -25.11),
 (-0.0, 'CH8_0669', 1.042253380736947, -25.11),
 (-0.0, 'CH8_0678', 1.042253380736947, -25.11),
 (-0.0, 'CH8_1804', 1.0422533807369174, -25.11),
 (-0.0, 'CH8_1805', 1.0422533807369174, -25.11),
 (-0.0, 'CH8_1821', 1.0422533807369174, -25.11),
 (-0.0, 'CH8_0703', 1.042253380736863, -25.11),
 (-0.0, 'CH8_1084', 1.0422533807368348, -25.11),
 (-0.0, 'CH8_0058', 1.0422533807367786, -25.11),
 (-0.0, 'CH8_1109', 1.042253380736776, -25.11),
 (-0.0, 'CH8_1107', 1.042253380736776, -25.11),
 (-0.0, 'CH8_1108', 1.042253380736776, -25.11),
 (-0.0, 'CH8_1106', 1.042253380736776, -25.11),
 (-0.0, 'CH8_0572', 1.042253380736776, -25.11),
 (-0.0, 'CH8_1458', 1.042253380736776, -25.11),
 (-0.0, 'CH8_1988', 1.042253380736776, -25.11),
 (-0.0, 'CH8_0954', 1.042253380736776, -25.11),
 (-0.0, 'CH8_0211', 1.04225338073

In [111]:
for i in range(len(list_variants_D)):
    if list_variants_D[i][0] != float(0.0) and list_variants_D[i][2] != float(0.0):
        print(list_variants_D[i])

In [107]:
for i in range(len(growthList)):
    if growthList[i][0] != float(0.0) and growthList[i][2] != float(0.0):
        print(growthList[i])

The last two for cicles search for a tuple where, at the same time, the Diacetyl production and the growth performance are diferent of zero. In one of the elements of the project group both cicles gave values of -8.154166842571751e-17 and 9.814656165599831e-14, respectively. So, the calculation of BPCY is below:

In [116]:
BPCY =(abs(-8.154166842571751e-17)*abs(9.814656165599831e-14))/abs(-25.11)
BPCY

3.187190118549444e-31

The list was sorted by the 1st value (Diacetyl production) and by the growth performance. We can note several things:\
-as the growth performance is 0 for the 4 best values of Diacetyl production, the calculaton of BPCY would be also 0 1/h.\
-however, there is one hit where both values are not 0 (this was observed only in one of the project group element), for mutant with deleted gene 'CH8_2083'. The respective BPCY is about 3.19e-31 mmol prod/ mmol subs/ h, wich is almost zero. This means that the production of Diacetyl or the growth performance is  almost zero too, and that is what we can see in list_variants_D. 

## Milk medium

In [96]:
model_milk = read_sbml_model("sth_novo.xml")

In [97]:
model_milk.objective = 'EX_caspep_e'

model_milk.reactions.biomass_STR.lower_bound=0.64
model_milk.reactions.biomass_STR.upper_bound=0.64

#change the conditions of the solution
model_milk.reactions.EX_lac_L_e.lower_bound=30
model_milk.reactions.EX_lac_L_e.upper_bound=40.8

model_milk.reactions.EX_lcts_e.lower_bound= (-25.11)
model_milk.reactions.EX_lcts_e.upper_bound= (-22.6)

model_milk.reactions.EX_gal_e.lower_bound=0
model_milk.reactions.EX_gal_e.upper_bound=1000

model_milk.reactions.EX_glc_D_e.lower_bound=2
model_milk.reactions.EX_glc_D_e.upper_bound=2.5

model_milk.reactions.EX_caspep_e.lower_bound=-1000
model_milk.reactions.EX_caspep_e.upper_bound=1000

model_milk.reactions.EX_mpept_e.lower_bound=0
model_milk.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_milk.reactions.EX_ala_D_e.lower_bound = 0
model_milk.reactions.EX_ala_D_e.upper_bound = 1000
model_milk.reactions.EX_ala_L_e.lower_bound = 0
model_milk.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_milk.reactions.EX_arg_L_e.lower_bound = 0
model_milk.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_milk.reactions.EX_asn_L_e.lower_bound = 0
model_milk.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_milk.reactions.EX_asp_L_e.lower_bound = 0
model_milk.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_milk.reactions.EX_cys_L_e.lower_bound = 0
model_milk.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_milk.reactions.EX_gln_L_e.lower_bound = 0
model_milk.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_milk.reactions.EX_glu_L_e.lower_bound=0
model_milk.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_milk.reactions.EX_gly_e.lower_bound=0
model_milk.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_milk.reactions.EX_his_L_e.lower_bound=0
model_milk.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_milk.reactions.EX_ile_L_e.lower_bound=0
model_milk.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_milk.reactions.EX_leu_L_e.lower_bound=0
model_milk.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_milk.reactions.EX_lys_L_e.lower_bound=0
model_milk.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_milk.reactions.EX_met_L_e.lower_bound=0
model_milk.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_milk.reactions.EX_phe_L_e.lower_bound=0
model_milk.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_milk.reactions.EX_pro_L_e.lower_bound=0
model_milk.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_milk.reactions.EX_ser_D_e.lower_bound=0
model_milk.reactions.EX_ser_D_e.upper_bound=1000
model_milk.reactions.EX_ser_L_e.lower_bound=0
model_milk.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_milk.reactions.EX_thr_L_e.lower_bound=0
model_milk.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_milk.reactions.EX_trp_L_e.lower_bound=0
model_milk.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_milk.reactions.EX_tyr_L_e.lower_bound=0
model_milk.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_milk.reactions.EX_val_L_e.lower_bound=0
model_milk.reactions.EX_val_L_e.upper_bound=1000

#### 1 a) What is the wild-type production of the compound?

In [116]:
model_milk.summary()

In [117]:
model_milk.optimize()["EX_diact_e"]

1.3461399902041513

Conclusion: The wild-type production of the Diacetyl is 1.346 mmol/gDW/h. 

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [118]:
model_milk_fva=fva(model_milk,fraction_of_optimum=1)
model_milk_fva.loc["EX_diact_e"] 

minimum    0.00000
maximum    7.05441
Name: EX_diact_e, dtype: float64

Conclusion: As the minimum and maximum are very distant to each other, we can conclude that the solution is not robust.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [119]:
model_milk_20 = read_sbml_model("sth_novo.xml")

In [124]:
model_milk_20.objective = 'EX_caspep_e'

#change the conditions of the solution
model_milk_20.reactions.EX_lac_L_e.lower_bound=30
model_milk_20.reactions.EX_lac_L_e.upper_bound=40.8

model_milk_20.reactions.EX_lcts_e.lower_bound= (-25.11)
model_milk_20.reactions.EX_lcts_e.upper_bound= (-22.6)

model_milk_20.reactions.EX_gal_e.lower_bound=0
model_milk_20.reactions.EX_gal_e.upper_bound=1000

model_milk_20.reactions.EX_glc_D_e.lower_bound=2
model_milk_20.reactions.EX_glc_D_e.upper_bound=2.5

model_milk_20.reactions.EX_caspep_e.lower_bound=-1000
model_milk_20.reactions.EX_caspep_e.upper_bound=1000

model_milk_20.reactions.EX_mpept_e.lower_bound=0
model_milk_20.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_milk_20.reactions.EX_ala_D_e.lower_bound = 0
model_milk_20.reactions.EX_ala_D_e.upper_bound = 1000
model_milk_20.reactions.EX_ala_L_e.lower_bound = 0
model_milk_20.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_milk_20.reactions.EX_arg_L_e.lower_bound = 0
model_milk_20.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_milk_20.reactions.EX_asn_L_e.lower_bound = 0
model_milk_20.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_milk_20.reactions.EX_asp_L_e.lower_bound = 0
model_milk_20.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_milk_20.reactions.EX_cys_L_e.lower_bound = 0
model_milk_20.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_milk_20.reactions.EX_gln_L_e.lower_bound = 0
model_milk_20.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_milk_20.reactions.EX_glu_L_e.lower_bound=0
model_milk_20.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_milk_20.reactions.EX_gly_e.lower_bound=0
model_milk_20.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_milk_20.reactions.EX_his_L_e.lower_bound=0
model_milk_20.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_milk_20.reactions.EX_ile_L_e.lower_bound=0
model_milk_20.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_milk_20.reactions.EX_leu_L_e.lower_bound=0
model_milk_20.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_milk_20.reactions.EX_lys_L_e.lower_bound=0
model_milk_20.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_milk_20.reactions.EX_met_L_e.lower_bound=0
model_milk_20.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_milk_20.reactions.EX_phe_L_e.lower_bound=0
model_milk_20.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_milk_20.reactions.EX_pro_L_e.lower_bound=0
model_milk_20.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_milk_20.reactions.EX_ser_D_e.lower_bound=0
model_milk_20.reactions.EX_ser_D_e.upper_bound=1000
model_milk_20.reactions.EX_ser_L_e.lower_bound=0
model_milk_20.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_milk_20.reactions.EX_thr_L_e.lower_bound=0
model_milk_20.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_milk_20.reactions.EX_trp_L_e.lower_bound=0
model_milk_20.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_milk_20.reactions.EX_tyr_L_e.lower_bound=0
model_milk_20.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_milk_20.reactions.EX_val_L_e.lower_bound=0
model_milk_20.reactions.EX_val_L_e.upper_bound=1000

In [125]:
model_milk_20.reactions.biomass_STR.lower_bound=0.64*0.2
model_milk_20.reactions.biomass_STR.upper_bound=0.64*0.2 

In [127]:
model_milk_20.optimize()["EX_diact_e"] 

-0.0

Conclusion: The Diacetyl production is 0.0 mmol/gDW/h.

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.

In [130]:
#saving Diacetyl production, gene name, growth performance and Lactose consumption in a list
list_variants_D_MM=[]
for i in model_milk.genes:
    with model_milk:
        variant=()
        objective,growth=0,0
        gene=''
        i.knock_out()
        tmp=model_milk.optimize()
        save_diact_optimize=float(tmp["EX_diact_e"])
        save_lactose_optimize=float(tmp["EX_lcts_e"])
        gene=str(i)
        objective=model_milk.optimize()['biomass_STR']
        variant = (save_diact_optimize, gene, objective,save_lactose_optimize)
        list_variants_D_MM.append(variant)

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

In [131]:
#ranking in order of Diacetyl production
list_variants_D_MM_sorted=sorted(list_variants_D_MM, reverse=True)
list_variants_D_MM_sorted

[(7.172105439999994, 'CH8_0716', 0.64, -24.958192693333476),
 (7.172105439999994, 'CH8_0715', 0.64, -24.958192693333476),
 (6.603964540869921, 'CH8_0509', 0.64, -25.11),
 (6.603964540869921, 'CH8_0503', 0.64, -25.11),
 (6.603964540869921, 'CH8_0493', 0.64, -25.11),
 (6.389491223576737, 'CH8_0812', 0.64, -24.24120540743633),
 (5.245146960000852, 'CH8_0428', 0.64, -25.11),
 (5.245146960000852, 'CH8_0423', 0.64, -25.11),
 (5.214985920001405, 'CH8_0352', 0.64, -25.11),
 (5.2149859200002755, 'CH8_0368', 0.64, -25.11),
 (5.150485635404582, 'CH8_0483', 0.64, -24.19760763937634),
 (5.149998591999669, 'CH8_0782', 0.64, -23.20558856533314),
 (5.149998591999669, 'CH8_0717', 0.64, -23.20558856533314),
 (5.149998591999669, 'CH8_0714', 0.64, -23.20558856533314),
 (5.149998591999669, 'CH8_0712', 0.64, -23.20558856533314),
 (4.697478288002795, 'CH8_0473', 0.64, -24.588912864001895),
 (4.697478288002795, 'CH8_0464', 0.64, -24.588912864001895),
 (4.697478288002795, 'CH8_0455', 0.64, -24.588912864001895)

In [132]:
BPCY=0
for i in range(0,4):
    BPCY=(abs(list_variants_D_MM_sorted[i][0])*abs(list_variants_D_MM_sorted[i][2]))/abs(list_variants_D_MM_sorted[i][3])
    print(BPCY,'mmol prod /mmol subs/h')

0.18391345631472986 mmol prod /mmol subs/h
0.18391345631472986 mmol prod /mmol subs/h
0.1683208803726304 mmol prod /mmol subs/h
0.1683208803726304 mmol prod /mmol subs/h


In order to simplify the analysis of the results, we calculated the BPCY only for the first 4 values (last script), where the growth performance is fixed to 0.64 1/h.\
We could see that the higher production of the Diacetyl corresponds to about 7.172 mmol/gDW/h, with about 24.958 mmol/gDW/h lactose consumption, so the higher BPCY is about 0.1839 mmol product /mmol subs /h. This means that, per hour of growth, per mmol of lactose consumption, 0.1839 mmol of Diacetyl are being produced, so the consuption of lactose is higher than the Diacetyl production (per hour of growth).

In [141]:
## EXTRA
#ranking in order of Diacetyl production and objective function: minimization of casein consumption
list_variants_D_MM_cas=[]
for i in model_milk.genes:
    with model_milk:
        variant=()
        objective,growth=0,0
        gene=''
        i.knock_out()
        tmp=model_milk.optimize()
        save_diact_optimize=float(tmp["EX_diact_e"])
        save_lactose_optimize=float(tmp["EX_lcts_e"])
        gene=str(i)
        objective=model_milk.slim_optimize()
        variant = (save_diact_optimize, gene, objective,save_lactose_optimize)
        list_variants_D_MM_cas.append(variant)

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\guilh\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\U

In [142]:
#the deletion of some genes turns the system infeasable -> "nan", so this script deletes those
import numpy as np
cleanedListMM_cas = []
for y in list_variants_D_MM_cas:
    if not np.isnan(y[2]):
        cleanedListMM_cas.append(y)

In [143]:
#sorting in order of Diacetyl production and minimization of Caseíne consumption
def Sort(sub_li):
    sub_li.sort(key = lambda x: x[0] and x[2], reverse=True)
    return sub_li
growthListMM_cas=Sort(cleanedListMM_cas)

In [144]:
for i in range(len(growthListMM_cas)):
    if growthListMM_cas[i][0] != float(0.0) and growthListMM_cas[i][2] != float(0.0):
        print(growthListMM_cas[i])

(5.593578872381098, 'CH8_1375', -0.0030233599999999537, -25.11)
(3.594615014557017, 'CH8_1376', -0.0030233599999999537, -25.11)
(3.594615014557017, 'CH8_1390', -0.0030233599999999537, -25.11)
(3.594615014557017, 'CH8_1411', -0.0030233599999999537, -25.11)
(6.226235520000081, 'CH8_0581', -0.003023359999999994, -25.11)
(4.795823573332957, 'CH8_0517', -0.003023359999999994, -25.11)
(4.795823573332957, 'CH8_0518', -0.003023359999999994, -25.11)
(4.795823573332957, 'CH8_0519', -0.003023359999999994, -25.11)
(4.795823573332957, 'CH8_0520', -0.003023359999999994, -25.11)
(4.795823573332957, 'CH8_0521', -0.003023359999999994, -25.11)
(4.795823573332957, 'CH8_0529', -0.003023359999999994, -25.11)
(4.795823573332957, 'CH8_0530', -0.003023359999999994, -25.11)
(6.510799146666788, 'CH8_1346', -0.003023359999999994, -25.11)
(6.510799146666788, 'CH8_1347', -0.003023359999999994, -25.11)
(6.510799146666788, 'CH8_1348', -0.003023359999999994, -25.11)
(3.5461312495483615, 'CH8_1545', -0.003023359999999

We did a new list (A,B,C,D), where:\
A - Diacetyl production\
B - deleted gene\
C - objective function: minimization of casein consumption\
D - lactose consumption

This was an extra excercise, only to visually analyse the ranking of the mutants when comparing the production of the Diacetyl with the objective function. And here we don't calcule the BPCY.
We can see that the production of Diacetyl can be enhanced when specific genes are deleted.
So, for example, when the gene 'CH8_0247' is deleted, a production of about 5.59 mmol/gDW/h of Diacetyl occurs at the same time that the model has the same value of objective function as before of any mutation.

#### 2) The authors of the model obtained gene expression data (RNA-seq) for this strain in CDM and milk medium (“gene_expression.xlsx”) in anaerobic conditions. For this exercise, consider the maximization of the growth rate as the objective for both media. For milk medium, set the uptake of casein as 0.004 mmol/gDW/h.

#### a) Integrate the expression data for both conditions using the GIMME algorithm. Compare the production of your compound in both conditions, as well as the flux of its production pathway(s).

In [3]:
gene_expression = pandas.read_excel("gene_expression.xlsx")

## GIMME- Chemical defined medium (CDM)

In [89]:
envcond = {
"EX_o2_e":(0,0) 
 }
#this change the medium to anaerobic one

simul = get_simulator(model, envcond = envcond)

#get genes and expression data
genes = gene_expression.ID.values  
conditions = ["CDM"]
exp = gene_expression[conditions].to_numpy()

# load expression module
expr = ExpressionSet(genes, conditions, exp)

# Apply parsimonious GIMME in the model
g_CDM = GIMME(simul, expr, parsimonious=True)

print(g_CDM)
print()
print(g_CDM.dataframe.loc["EX_diact_e"])
print(g_CDM.dataframe.loc["EX_lcts_e"])
print(g_CDM.dataframe.loc["biomass_STR"])

objective: biomass_STR
Status: OPTIMAL
Constraints: OrderedDict([('EX_o2_e', (0, 0)), ('biomass_STR', (0.9380280426627976, inf))])
Method:None

Flux rate   -0.0
Name: EX_diact_e, dtype: float64
Flux rate   -25.11
Name: EX_lcts_e, dtype: float64
Flux rate    0.938028
Name: biomass_STR, dtype: float64


In [91]:
#calculation of BPCY with GIMME simulation
BPCY =abs(g_CDM.dataframe.loc["EX_diact_e"])*abs(g_CDM.dataframe.loc["biomass_STR"])/abs(g_CDM.dataframe.loc["EX_lcts_e"])
BPCY

Flux rate    0.0
dtype: float64

In [93]:
#calculation of BPCY with FBA-model (just for curiosity)
model.summary()
BPCY=abs(model.optimize()["EX_diact_e"])*abs(model.slim_optimize())/abs(model.optimize()["EX_lcts_e"])
BPCY

0.0

## GIMME- Milk medium (MM)

In [30]:
model_GIMME_MM = read_sbml_model("sth_novo.xml")


#change the conditions of the solution
model_GIMME_MM.reactions.EX_lac_L_e.lower_bound=30
model_GIMME_MM.reactions.EX_lac_L_e.upper_bound=40.8

model_GIMME_MM.reactions.EX_lcts_e.lower_bound= (-25.11)
model_GIMME_MM.reactions.EX_lcts_e.upper_bound= (-22.6)

model_GIMME_MM.reactions.EX_gal_e.lower_bound=0
model_GIMME_MM.reactions.EX_gal_e.upper_bound=1000

model_GIMME_MM.reactions.EX_glc_D_e.lower_bound=2
model_GIMME_MM.reactions.EX_glc_D_e.upper_bound=2.5

model_GIMME_MM.reactions.EX_caspep_e.lower_bound=-0.004
#model_GIMME_MM.reactions.EX_caspep_e.upper_bound=1000

model_GIMME_MM.reactions.EX_mpept_e.lower_bound=0
model_GIMME_MM.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_GIMME_MM.reactions.EX_ala_D_e.lower_bound = 0
model_GIMME_MM.reactions.EX_ala_D_e.upper_bound = 1000
model_GIMME_MM.reactions.EX_ala_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_GIMME_MM.reactions.EX_arg_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_GIMME_MM.reactions.EX_asn_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_GIMME_MM.reactions.EX_asp_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_GIMME_MM.reactions.EX_cys_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_GIMME_MM.reactions.EX_gln_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_GIMME_MM.reactions.EX_glu_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_GIMME_MM.reactions.EX_gly_e.lower_bound=0
model_GIMME_MM.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_GIMME_MM.reactions.EX_his_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_GIMME_MM.reactions.EX_ile_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_GIMME_MM.reactions.EX_leu_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_GIMME_MM.reactions.EX_lys_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_GIMME_MM.reactions.EX_met_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_GIMME_MM.reactions.EX_phe_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_GIMME_MM.reactions.EX_pro_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_GIMME_MM.reactions.EX_ser_D_e.lower_bound=0
model_GIMME_MM.reactions.EX_ser_D_e.upper_bound=1000
model_GIMME_MM.reactions.EX_ser_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_GIMME_MM.reactions.EX_thr_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_GIMME_MM.reactions.EX_trp_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_GIMME_MM.reactions.EX_tyr_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_GIMME_MM.reactions.EX_val_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_val_L_e.upper_bound=1000

In [108]:
envcond = {
"EX_o2_e":(0,0) #oxygen exchange
 }

simul = get_simulator(model_GIMME_MM, envcond = envcond)

#get genes and expression data
genes = gene_expression.ID.values  
conditions = ["MILK"]
exp = gene_expression[conditions].to_numpy()

# load expression module
expr = ExpressionSet(genes, conditions, exp)

# Apply parsimonious GIMME in the model
g = GIMME(simul, expr, parsimonious=True)
print(g)
print()
print(g.dataframe.loc["EX_diact_e"])
print(g.dataframe.loc["EX_lcts_e"])
print(g.dataframe.loc["biomass_STR"])

objective: biomass_STR
Status: OPTIMAL
Constraints: OrderedDict([('EX_o2_e', (0, 0)), ('biomass_STR', (0.7620660457239588, inf))])
Method:None

Flux rate    2.57237
Name: EX_diact_e, dtype: float64
Flux rate   -22.6
Name: EX_lcts_e, dtype: float64
Flux rate    0.762066
Name: biomass_STR, dtype: float64


In [88]:
#calculation of BPCY with GIMME simulation
BPCY =abs(g.dataframe.loc["EX_diact_e"])*abs(g.dataframe.loc["biomass_STR"])/abs(g.dataframe.loc["EX_lcts_e"])
BPCY

Flux rate    0.08674
dtype: float64

In [103]:
#calculation of BPCY with FBA-model_milk (just for curiosity)
model_milk.summary()
BPCY=abs(model_milk.optimize()["EX_diact_e"])*0.64/abs(model_milk.optimize()["EX_lcts_e"])
print(BPCY)
print(model_milk.optimize()["EX_lcts_e"])
print(model_milk.optimize()["EX_diact_e"])

0.03812077848365738
-22.6
1.3461399902041513


Comparing the production of Diacetyl in both media, wich is 0 mmol/gDW/h for CDM medium and 2.573 mmol/gDW/h for milk medium, the BPCY will be 0 and 0.0867 mmol prod/mmol subs/h, respectively. The milk medium is more favorable for the Diacetyl production, growth performance and it consumes less substrate (Lactose). 

#### b) Compare the output of the algorithm with the results obtained in exercise 1 a).

Comparing GIMME with the previous model FBA, we can conclude that the current model can be more realistic as we can introduce the expression of genes from the real data of the organism that is being studied.\
So, when we compare the CDM medium, after introducing more details to the GIMME model:\
-The growth performance was lower (0.938 1/h) than the FBA one (1.042 1/h).\
-The Diacetyl production in both simulations is 0.0 mmol/gDW/h.\
And,when we compare the milk medium, after introducing more details to the GIMME model:\
-The growth performance was higher (0.762 1/h) than the FBA one (0.64 1/h). However, the growth performance in FBA simulation was locked to 0.64 1/h according to the exercise instructions.\
-The Diacetyl production was higher (2.572 mmol/gDW/h) than the FBA one (1.346 mmol/gDW/h).\
In conclusion, GIMME simulation gives different results from FBA as it deals with more data (expression of genes). If, in the future, we would have a project with gene expression data available, it would be better to use the GIMME simulation.